File: 307.区域和检索 - 数组可修改/NumArray.go
package problem0307

// 线段树的实现
type SegmentTree struct {
	start    int
	end      int
	property int
	left     *SegmentTree
	right    *SegmentTree
}

func buildSegmentTree(start int, end int, nums []int) *SegmentTree {
	if start > end {
		return nil
	}
	if start == end {
		return &SegmentTree{
			start:    start,
			end:      end,
			property: nums[start],
			left:     nil,
			right:    nil,
		}
	}
	mid := (start + end) >> 1
	left := buildSegmentTree(start, mid, nums)
	right := buildSegmentTree(mid+1, end, nums)
	return &SegmentTree{
		start:    start,
		end:      end,
		property: left.property + right.property,
		left:     left,
		right:    right,
	}
}

func (this *SegmentTree) updateSegmentTree(i int, val int) {
	if this.start == i && this.end == i {
		this.property = val
		return
	}
	mid := (this.start + this.end) >> 1
	if i <= mid {
		this.left.updateSegmentTree(i, val)
	} else {
		this.right.updateSegmentTree(i, val)
	}
	this.property = this.left.property + this.right.property
}

func (this *SegmentTree) rangeSegmentTree(i int, j int) int {
	if this.start == i && this.end == j {
		return this.property
	}
	mid := (this.start + this.end) >> 1
	if j <= mid {
		return this.left.rangeSegmentTree(i, j)
	} else if i > mid {
		return this.right.rangeSegmentTree(i, j)
	} else {
		return this.left.rangeSegmentTree(i, mid) + this.right.rangeSegmentTree(mid+1, j)
	}
}

// 给定一个整数数组nums，求出索引[i, j]范围内的和
// update(i, val)可以将下标为i的数值更新为val
type NumArray struct {
	segment *SegmentTree
}

func Constructor(nums []int) NumArray {
	return NumArray{
		segment: buildSegmentTree(0, len(nums)-1, nums),
	}
}

func (this *NumArray) Update(i int, val int) {
	this.segment.updateSegmentTree(i, val)
}

func (this *NumArray) SumRange(i int, j int) int {
	return this.segment.rangeSegmentTree(i, j)
}

File: 307.区域和检索 - 数组可修改/NumArray.js
class NumArray {
    constructor(nums) {
        this.nums = nums
        this.n = nums.length
        this.tree = new Array(4 * this.n)
        this._build(0, this.n - 1, 1)
    }
    // 初始化
    _build(start, end, i) {
        if (start === end) {
            this.tree[i] = this.nums[start]
            return
        }
        const mid = start + Math.floor((end - start) / 2)
        this._build(start, mid, i * 2)
        this._build(mid + 1, end, i * 2 + 1)
        this._pushUp(i)
    }
    _pushUp(i) {
        this.tree[i] = this.tree[i * 2] + this.tree[i * 2 + 1]
    }
    // 单点修改，分为：增量式修改（加上某值）、覆盖式修改（修改为某值）
    _update(updateI, val, start, end, i, isOverwrite) {
        if (start === end) {
            if (isOverwrite) {
                this.tree[i] = val
                this.nums[updateI] = val // 实时维护nums
            } else {
                this.tree[i] += val
                this.nums[updateI] += val
            }
            return
        }
        const mid = start + Math.floor((end - start) / 2)
        if (updateI <= mid) {
            this._update(updateI, val, start, mid, i * 2, isOverwrite)
        } else {
            this._update(updateI, val, mid + 1, end, i * 2 + 1, isOverwrite)
        }
        this._pushUp(i)
    }
    // 区间查询
    _sum(left, right, start, end, i) {
        if (left <= start && end <= right) {
            return this.tree[i]
        }
        const mid = start + Math.floor((end - start) / 2)
        let sum = 0
        if (left <= mid) {
            sum += this._sum(left, right, start, mid, i * 2)
        }
        if (right > mid) {
            sum += this._sum(left, right, mid + 1, end, i * 2 + 1)
        }
        return sum
    }
    // 区间极值
    _minOrMax(left, right, start, end, i, isMIn) {
        if (start === end) {
            return this.tree[i]
        }
        const mid = start + Math.floor((end - start) / 2)
        let lValue = isMIn ? Number.MAX_VALUE : Number.MIN_VALUE
        let rValue = isMIn ? Number.MAX_VALUE : Number.MIN_VALUE
        if (left <= mid) {
            lValue = this._minOrMax(left, right, start, mid, i * 2, isMIn)
        }
        if (right > mid) {
            rValue = this._minOrMax(left, right, mid + 1, end, i * 2 + 1, isMIn)
        }
        return Math[isMIn ? 'min' : 'max'](lValue, rValue)
    }
    add(updateI, val) {
        this._update(updateI, val, 0, this.n - 1, 1, false)
    }
    update(updateI, val) {
        this._update(updateI, val, 0, this.n - 1, 1, true)
    }
    sumRange(left, right) {
        return this._sum(left, right, 0, this.n - 1, 1)
    }
    min(left, right) {
        return this._minOrMax(left, right, 0, this.n - 1, 1, true)
    }
    max(left, right) {
        return this._minOrMax(left, right, 0, this.n - 1, 1, false)
    }
}